src/main/java/com/facebook/openwifirrm/modules/ModelerUtils.java

```diff
@@ -202,4 +208,73 @@ public static double calculateTPCMetrics(
 			return Double.POSITIVE_INFINITY;
 		}
 	}
+
+	/**
+	 *
+	 * @param dataModel
+	 * @param obsoletionPeriodMs the maximum amount of time (in milliseconds) it is
+	 *                           worth aggregating over, starting from the most
+	 *                           recent scan entry and working backwards in time
+	 * @param agg                an aggregator to output a calculated "aggregated
+	 *                           signal strength"
+	 * @return a map from BSSID to an "aggregated wifiscan entry" (i.e., a single
+	 *         entry with its {@code signal} attribute modified to be the aggregated
+	 *         signal value instead of the value in just the most recent entry.
+	 */
+	public static Map<String, ProcessedWifiScanEntry> getAggregatedWifiScans(Modeler.DataModel dataModel,
+			long obsoletionPeriodMs,
+			Aggregator<Double> agg) {
+		/*
+		 * NOTE: if a BSSID does not have an entry, it will not be returned
+		 * (i.e., it will not be a key in the returned map).
+		 */
+		Map<String, ProcessedWifiScanEntry> aggregatedWifiScans = new HashMap<>();
+		for (Map.Entry<String, List<List<ProcessedWifiScanEntry>>> mapEntry : dataModel.latestWifiScans.entrySet()) {
+			// Flatten the wifiscan entries and sort in reverse chronological order
+			List<List<ProcessedWifiScanEntry>> scans = mapEntry.getValue();
+			if (scans.isEmpty()) {
+				continue;
+			}
+			List<ProcessedWifiScanEntry> mostRecentToOldest = scans.stream().flatMap(list -> list.stream())
+					.sorted((entry1, entry2) -> {
+						return -Long.compare(entry1.unixTimeMs, entry2.unixTimeMs);
+					}).collect(Collectors.toUnmodifiableList());
+
+			/*
+			 * For a given BSSID, discard the obsolete entries. Then,consider the most
+			 * recent entry. Take its ht_oper and vht_oper. Discard earlier entries with a
+			 * different ht_oper or a different vht_oper. For earlier entries with the same
+			 * ht_oper and vht_oper, take the average signal value.
+			 */
+			String bssid = mapEntry.getKey();
+			long now = mostRecentToOldest.get(0).unixTimeMs;
+			String newestHtOper = null;
+			String newestVhtOper = null;
+			agg.reset();
+			for (ProcessedWifiScanEntry entry : mostRecentToOldest) {
+				if (now - entry.unixTimeMs >= obsoletionPeriodMs) {
+					// discard obsolete entries
+					break;
+				} else if (entry.ht_oper == null && entry.vht_oper == null) {
+					continue;
+				}
+				if (newestHtOper == null) {
+					// may still be null after this assignment
+					newestHtOper = entry.ht_oper;
+				}
+				if (newestVhtOper == null) {
+					// may still be null after this assignment
+					newestVhtOper = entry.vht_oper;
+				}
+				// if the entry matches ht_oper and vht_oper, add its signal to the aggregate
+				if ((entry.ht_oper == null || entry.ht_oper == newestHtOper)
+						&& (entry.vht_oper == null || entry.vht_oper == newestVhtOper)) {
+					aggregatedWifiScans.putIfAbsent(bssid, new ProcessedWifiScanEntry(entry));
+					agg.addValue((double) entry.signal);
+				}
+			}
```

Comment:
I got some time to read through this. :) I think one case might be missing in the current implementation. 

If the most recent wifiscan has ht_oper = null and vht_oper = null, then you need to check the second most recent one to see whether it is still ht_oper = null and vht_oper = null. The reason is because its firmware doesn't support the wifiscan feature but we still need to aggregate them. In the current logic, it will skip this case in line 259. 

I believe the easiest way to do this is to assign newestHtOper =  mostRecentToOldest.get(0).ht_oper and newestVhtOper = mostRecentToOldest.get(0).vht_oper and then check the delta between the most recent one and the remaining entires. 

Again, I might be wrong :). Just share what I think. Please ignore my comment if you think it doesn't make sense. I will remove the change request in this PR so you won't be blocked. Thanks. :)

Reply:
Interesting, I did not realize this. What does it mean for the firmware to not support the scan? If that were the case, why would we have a `WifiScanEntry` at all? And in this case, does it make sense to aggregate, if we don't know the channel info? (I'm assuming if it doesn't support a scan, the other fields would be invalid as well)

Reply:
Oh sorry that I should have been more precise. “The wifiscan feature” here means a valid wifiscan but not reporting those two fields (ht_oper and vht_oper). In this case, we still have the rest of things (including channel, signal, etc.,) but not those two. This happens when someone use the older (but still valid) version of firmware, i.e., <2.4 (those two fields are supported after 2.5). 

Again, I could be wrong :) so please feel free to skip this comment. Just share what I have seen in the data before.

Reply:
Got it. What you're saying makes sense. In that case, would we want to do some kind of checking before aggregating still? E.g., check that `channel` and `frequency` are the same and aggregate only entries with the same channel and frequency?

Reply:
Those two are good IMHO. Thanks. :)

Reply:
I'll work on this in the separate PR, thank you!